### ISSUE TITLE
Possible better keyboard layout

### ISSUE THREAD

**Issue by rjp** [Hardmath123/hardmath123.github.io#22]
I've been trying various methods to see if I can find a better layout than `ZKVGWCDNIAHTESROLUMFYBPXJQ`. Unfortunately, my costs are slightly different than yours -- I got the three novels from Gutenberg and trimmed the Gutenberg header/footer before concatenating and stripping non-alphas. For example, you get a score of 5,499,341 whereas mine gets 5,262,321 for your best keyboard.

After many attempts at various "clever" solutions, a random search finally got something.

```2019-01-07 17:20:48.614733500 QJXPBYFUMLORSETHAINDWCGVKZ =5261961```

But since I don't get the scores as you, I can't really verify if it's "better"...

Could you possibly check against your input file to see if they're any improvement?

ANNOYINGLY, whilst typing this, my random search has found a better one - just one swap away from yours (`CW` vs `WC`):
```2019-01-07 19:25:32.725170500 ZKVGCWDNIAHTESROLUMFYBPXJQ =5260927
```

Thanks! And thanks for an excellent puzzle. It has been most absorbing the last few days.

**Comment by kach** [comment]
Ahh!! I'm so glad you found this interesting!

Okay, I ran "QJXPBYFUMLORSETHAINDWCGVKZ" on my program and got the score of `5501433`. It's likely because I preserved the Gutenberg header/footer. I doubt it makes a big difference in the frequency distribution, but it's just the extra characters that inflate my scores.

You can play with my code — it's actually in this repository! — at https://github.com/Hardmath123/hardmath123.github.io/tree/master/static/crown-typewriter. fast.cc is the one I used, and `digraphs.h` is the frequency distribution database as generated by `scrape-cpp.py`. A caveat is that you need to compile it with pthreads, so try `g++ -std=gnu++11 -pthread fast.cc` or equivalent. (And ignore the scrape.py and digraphs.json file, those were an earlier attempt in Python that I decided was too slow… and it uses only one novel to build its database).

**Comment by kach** [comment]
Good luck! I'm very very curious to see if there's a better permutation — please keep me posted. :-)

**Comment by BenjaminPoilve** [comment]
For anyone that is interested, I reproduced the method in a Notebook! Would like to try constraints-based solver on that (as suggested in the post!). Fun article by the way, reminded me of Peter Norvig's pytudes!

**Comment by rjp** [comment]
I’ve got a blog post coming about all the solutions I’ve tried including a Z3 solution as per the original blog post. I’m a complete Z3 neophyte, mind, but it hasn’t found any good solutions yet - they’re all above 7M (IIRC) no matter what I try. I suspect the search space is just too big...

**Comment by BenjaminPoilve** [comment]
Z3 indeed does not seem to be able to calculate a solution in a timely manner!

Here is the output from python, trying to calculate for a subset of letter, adding one at a time:
> 2  letters, takes 0.013633012771606445 seconds to compute
> 3  letters, takes 0.041387081146240234 seconds to compute
> 4  letters, takes 0.18227195739746094 seconds to compute
> 5  letters, takes 0.7077188491821289 seconds to compute
> 6  letters, takes 3.8231089115142822 seconds to compute
> 7  letters, takes 47.89865016937256 seconds to compute

I don't like that progression! Back of the envelope tells me the time it will take will be counted in millennium!

**Comment by rjp** [comment]
Yeah, I left my 26 node one running for two hours and it didn’t even produce a good solution.

**Comment by rjp** [comment]
Part 1 (because it was taking forever and I still have loads to document)

http://rjp.is/blogging/posts/2019/01/linear-typewriters/

**Comment by kach** [comment]
Wow! That's amazing — so many strategies I hadn't even thought of. I'm going to post a link to that as an update to the original piece. =)

It's interesting that you mention that it's also hard to _verify_ a purported optimal solution. I wonder if that puts us in coNP territory…
